### ISSUE TITLE
Is it useful to cache calls to SqlDataReader.GetOrdinal, or is that already done internally?

### ISSUE THREAD

**Issue by cmeeren** [dotnet/SqlClient#837]
Fetching a `SqlDataReader` column value by indexing on the column name causes a call to `GetOrdinal`:

https://github.com/dotnet/SqlClient/blob/6a9e4f104ce77ba631e2ac8ea2e366a4d41ba235/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs#L426-L432

It may look like `GetOrdinal` uses some kind of caching internally:

https://github.com/dotnet/SqlClient/blob/6a9e4f104ce77ba631e2ac8ea2e366a4d41ba235/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDataReader.cs#L1416-L1421

I have tried to follow that trail, but I have not been able to ascertain whether there is caching involved in this lookup.

Can it be worthwile, performance-wise, to implement my own caching of column names to ordinals when fetching the result rows of a query, and just use ordinal-based lookup based on the cached values? Or will that be superfluous since it seems ordinal lookups may already be cached in SqlDataReader when using name-based lookups?

**Comment by JRahnama** [comment]
@cmeeren Thanks for bringing this up to our attention. `FieldNameLookup` passes the reader to its `base` which is `BasicFieldNameLookup` and that creates a `string[]` with names of the fieldNames stored in it. The names are stored in that array for the life cycle of the reader. Does this answer your question?

**Comment by Wraith2** [comment]
Yes the field lookups are cached internally as @JRahnama said in `FieldNameLookup` but if you want to run as fast as possible it's a reasonable thing to do to cache the field name ordinal once and then re-use it, you avoid the lookup instructions and possible allocations for the map. I've got a prototype of this functionality that uses spans and temp char buffers that I'll get around to productizing eventually.

**Comment by cmeeren** [comment]
> Yes the field lookups are cached internally as @JRahnama said in `FieldNameLookup` but if you want to run as fast as possible it's a reasonable thing to do to cache the field name ordinal once and then re-use it, you avoid the lookup instructions and possible allocations for the map.

I'm confused; the lookups are cached, but it's still reasonable to cache them ourselves? Does that mean that it's inefficiently cached?

> I've got a prototype of this functionality that uses spans and temp char buffers that I'll get around to productizing eventually.

Looking forward to it.

**Comment by Wraith2** [comment]
Even a cached string lookup takes a finite amount of time to do the lookup. The fastest way to access is directly by ordinal. On the other hand since you're doing network IO to get data from the sql server it's likely you'll end up waiting on that anyway. Don't rely on intuition of this sort of thing use a profiler to identify where you're spending time, the odds are very high that it isn't where you think it is.

**Comment by cmeeren** [comment]
Thanks!

> Don't rely on intuition of this sort of thing use a profiler to identify where you're spending time, the odds are very high that it isn't where you think it is.

My problem here is that I am creating a general data access library (in the form of a build-time code generator), so I have no concrete workloads to test against.

I could of course make the assumption that the worst-case scenario in terms of `GetOrdinal` overhead would be a query that fetches lots of rows (i.e. lots of calls to `GetOrdinal` that could be cached after the first row) with narrow columns (to minimize the IO of reading the rows). I would expect the number of columns to not matter that much, since `GetOrdinal` is called once per column. Would that be a case that makes sense to test against?

Otherwise, if you have any insights in terms of the scenario described above, I'm all ears.

**Comment by JRahnama** [comment]
@cmeeren if your question is answered and there is nothing more to be done here feel free to close the issue.

Thanks

**Comment by cmeeren** [comment]
Closing. For the record, I solved this by generating one `int` field for each (statically known) result set column, and when fetching the rows at runtime, each query calls `GetOrdinal` for the column names and sets these `int` fields before reading the rows using the ordinals.
